src/function/scalar/date/time_bucket.cpp

```diff
@@ -0,0 +1,386 @@
+				}
+				return OffsetWidthMoreThanMonthsTernaryOperator::Operation<TA, TB, TC, TR>(bucket_width, ts, offset);
+			} else {
+				throw NotImplementedException("Month intervals cannot have day or time component");
+			}
+		}
+	};
+
+	struct OriginWidthLessThanDaysTernaryOperator {
+		template <class TA, class TB, class TC, class TR>
+		static inline TR Operation(TA bucket_width, TB ts, TC origin) {
+			if (!Value::IsFinite(ts)) {
+				return Cast::template Operation<TB, TR>(ts);
+			}
+			int64_t bucket_width_micros = Interval::GetMicro(bucket_width);
+			int64_t ts_micros = Timestamp::GetEpochMicroSeconds(Cast::template Operation<TB, timestamp_t>(ts));
+			int64_t origin_micros = Timestamp::GetEpochMicroSeconds(Cast::template Operation<TB, timestamp_t>(origin));
+
+			return Cast::template Operation<timestamp_t, TR>(Timestamp::FromEpochMicroSeconds(
+			    WidthLessThanDaysCommon(bucket_width_micros, ts_micros, origin_micros)));
+		}
+	};
+
+	struct OriginWidthMoreThanMonthsTernaryOperator {
+		template <class TA, class TB, class TC, class TR>
+		static inline TR Operation(TA bucket_width, TB ts, TC origin) {
+			if (!Value::IsFinite(ts)) {
+				return Cast::template Operation<TB, TR>(ts);
+			}
+			int32_t ts_months = EpochMonths(ts);
+			int32_t origin_months = EpochMonths(origin);
+
+			int32_t result_months = WidthMoreThanMonthsCommon(bucket_width.months, ts_months, origin_months);
+			int32_t year = ExtractYearFromEpochMonths(result_months);
+			int32_t month = ExtractMonthFromEpochMonths(result_months);
+			return Cast::template Operation<date_t, TR>(Date::FromDate(year, month, 1));
+		}
+	};
+
+	struct OriginTernaryOperator {
+		template <class TA, class TB, class TC, class TR>
+		static inline TR Operation(TA bucket_width, TB ts, TC origin, ValidityMask &mask, idx_t idx) {
+			if (!Value::IsFinite(origin)) {
+				mask.SetInvalid(idx);
+				return TR();
+			}
+			if (bucket_width.months == 0) {
+				int64_t bucket_width_micros = Interval::GetMicro(bucket_width);
+				if (bucket_width_micros <= 0) {
+					throw NotImplementedException("Period must be greater than 0");
+				}
+				return OriginWidthLessThanDaysTernaryOperator::Operation<TA, TB, TC, TR>(bucket_width, ts, origin);
+			} else if (bucket_width.months != 0 && bucket_width.days == 0 && bucket_width.micros == 0) {
+				if (bucket_width.months < 0) {
+					throw NotImplementedException("Period must be greater than 0");
+				}
+				return OriginWidthMoreThanMonthsTernaryOperator::Operation<TA, TB, TC, TR>(bucket_width, ts, origin);
+			} else {
+				throw NotImplementedException("Month intervals cannot have day or time component");
+			}
+		}
+	};
+};
+
+template <typename T>
+static void TimeBucketFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	D_ASSERT(args.ColumnCount() == 2);
+
+	auto &bucket_width_arg = args.data[0];
+	auto &ts_arg = args.data[1];
+
+	if (bucket_width_arg.GetVectorType() == VectorType::CONSTANT_VECTOR) {
```

Comment:
Is the point here that you want to throw for negative bucket sizes even if the other arguments are `NULL`?

Reply:
It seems possible to avoid throwing error by calling `BinaryExecutor:Execute`. I'll do so.